Mathlib/GroupTheory/MaximalSubgroups.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2024 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Algebra.Group.Subgroup.Lattice
+import Mathlib.Order.Atoms
+
+/-! # Maximal subgroups
+
+A subgroup `IsMaximal` if it is maximal in the collection of
+proper subgroups.
+
+We prove a few basic results which are essentially copied from
+those about maximal ideals.
+
+Besides them, we have :
+* `isMaximal_iff` proves that a subgroup K of G is maximal
+if and only  if it is not ⊤ and if for all g ∈ G such that g ∉ K,
+every subgroup containing K and g is ⊤.
+
+## TODO
+
+Is it useful to write it in a greater generality?
+Will we need to write this for all `sub`-structures ?
+
+-/
+
+
+variable {G : Type*} [Group G]
+
+namespace Subgroup
+
+/-- A subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class _root_.AddSubgroup.IsMaximal {G : Type*} [AddGroup G] (K : AddSubgroup G) : Prop where
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+  out : IsCoatom K
+
+/-- A subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+@[to_additive]
+class IsMaximal (K : Subgroup G) : Prop where
+/-- A subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+  out : IsCoatom K
+
+@[to_additive]
+theorem isMaximal_def {K : Subgroup G} : K.IsMaximal ↔ IsCoatom K :=
+  ⟨fun h => h.1, fun h => ⟨h⟩⟩
+
+@[to_additive]
+theorem IsMaximal.ne_top {K : Subgroup G} (h : K.IsMaximal) : K ≠ ⊤ :=
+  (isMaximal_def.1 h).1
+
+@[to_additive]
+theorem isMaximal_iff {K : Subgroup G} :
+    K.IsMaximal ↔ K ≠ ⊤ ∧ ∀ (H : Subgroup G) (g), K ≤ H → g ∉ K → g ∈ H → H = ⊤ := by
```

Comment:
I feel like we should make a `SetLike.isMaximal_iff` that unifies this and `Ideal.isMaximal_iff`.

Reply:
That seems to require that the substructure is a sup-lattice.

Reply:
And the proof is pretty easy:
```
import Mathlib.Data.SetLike.Basic
import Mathlib.Order.Atoms
variable (A B : Type*) [SetLike A B] [OrderTop A]
theorem SetLike.isCoatom_iff {K : A} :
    IsCoatom K ↔ K ≠ ⊤ ∧ ∀ H g, K ≤ H → g ∉ K → g ∈ H → H = ⊤ := by
  simp_rw [IsCoatom, lt_iff_le_not_le, SetLike.not_le_iff_exists,
    and_comm (a := _ ≤ _), and_imp, exists_imp, ← and_imp, and_comm (a := _ ≤ _), and_comm]
```